File: HW/HW7/HW7.py
from typing import List

class HW7:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        ## Catches improper input
        if not intervals:
            return "Incorrect formatting"
        
        ## Puts start and end into different arrays
        start_time = sorted(i[0] for i in intervals)
        end_time = sorted(i[1] for i in intervals)
        
        ## Initialize pointer
        in_use = 0
        start = end = 0
        intervals_size = len(intervals)
        
        ## Iterate through the intervals
        while start < intervals_size:
            if start_time[start] >= end_time[end]:
                
                ## Takes out a room
                in_use -= 1
                end += 1
            
            ## Room in use
            in_use += 1
            start += 1
        
        return in_use

hw7 = HW7()
intervals1 = [[0,30],[5,10],[15,20]]
intervals2 = [[0,1],[1,2],[2,3]]
print(hw7.minMeetingRooms(intervals1))
print(hw7.minMeetingRooms(intervals2))

File: HW/HW7/HW7.java
import java.util.Arrays;

public class HW7 {
    public int minMeetingRooms(int[][] intervals) {
        // Catches improper input
        if (intervals == null || intervals.length == 0)
            return 0; // or throw an exception for incorrect formatting
        
        // Puts start and end into different arrays
        int[] start_time = new int[intervals.length];
        int[] end_time = new int[intervals.length];
        for (int i = 0; i < intervals.length; i++) {
            start_time[i] = intervals[i][0];
            end_time[i] = intervals[i][1];
        }
        Arrays.sort(start_time);
        Arrays.sort(end_time);
        
        // Initialize pointers
        int in_use = 0;
        int start = 0;
        int end = 0;
        int intervals_size = intervals.length;
        
        // Iterate through the intervals
        while (start < intervals_size) {
            if (start_time[start] >= end_time[end]) {
                // Takes out a room
                in_use--;
                end++;
            }
            // Room in use
            in_use++;
            start++;
        }
        return in_use;
    }

    public static void main(String[] args) {
        HW7 hw7 = new HW7();
        int[][] intervals1 = {{0, 30}, {5, 10}, {15, 20}};
        int[][] intervals2 = {{0, 1}, {1, 2}, {2, 3}};
        System.out.println(hw7.minMeetingRooms(intervals1));
        System.out.println(hw7.minMeetingRooms(intervals2));
    }
}
